### DISCUSSION TITLE
syncing model_has_permissions with roles + permissions

### DISCUSSION THREAD

**Discussion by aa384** [spatie/laravel-permission#2114] [Q&A] [answered]
if I remove a permission from the role, the entry still exists in model_has_permissions. Is there a way that these can all be synced or will I have to do it manually? 

example - user 1 has role 1 with permission 1, I remove permission 1 from role 1 but user 1 is still left with permission 1

**Comment by erikn69** [comment, marked as answer]
> if I remove a permission from the role, the entry still exists in model_has_permissions.

role permision is handled by `role_has_permissions`, if you remove a permission from the role, the data on that table get deleted, not on `model_has_permissions`
https://github.com/spatie/laravel-permission/blob/fc841f86190f558658c072c2f2107fdf67308c8f/config/permission.php#L71

Maybe, after deleted you forget `$this->forgetCachedPermissions();` for cache refresh

**Reply by aa384** [reply]
so model_has_permissions is never called and I dont need to worry about the entries in there you are saying?

**Reply by erikn69** [reply]
Yes, it is never called on your example, but it is a good idea adding `forgetCachedPermissions` after `role` or `permission` delete

**Reply by erikn69** [reply]
No, i don't misunderstood your question. You misunderstood how package works, 

>if I have role number 1 with permission number 1 and 2. if I give user 1 , role 1. the model_has_permissions table will save an entry for user 1 with both permission number 1 and 2

I repeat again, when you assign a role number 1 to a user 1, there is not a entrie on `model_has_permissions`, it adds a entrie on `model_has_roles`, when you assign to role number 1 the permission number 1 and 2, there is two entries o `role_has_permissions`, `model_has_permissions` is never used on this example

https://github.com/spatie/laravel-permission/blob/dbc6c18ef4b2d097350a5faf82382564748e6c97/config/permission.php#L55-L71

```php
$role1->givePermissionTo(['permission1', 'permission2']); // add two entries on role_has_permissions
$user1->assignRole($role1); // add one entrie on model_has_roles
$role1->delete(); // remove 1 entrie on model_has_roles, and 2 entries on role_has_permissions
// model_has_permissions was never used
```

**Reply by aa384** [reply]
thank you @erikn69. I think I have explained my self poorly here, but I appreciate you taking the time to respond.

**Reply by erikn69** [reply]
Maybe you are want to delete permissions insted of roles, if you delete permission 1, it removes `model_has permissions` and `role_has_permissions` entries

**Comment by aa384** [comment]


**Reply by angeljqv** [reply]
Seriously, did you mark yourself as answerer without any answer, that is why people stop helping

**Reply by aa384** [reply]
@angeljqv I have updated my answer with more detail
